### PR TITLE
Bluetooth: host: cancel limited adv timeout when advertising stopped

### LINKED ISSUE #37467: Bluetooth: host: Incorrect advertiser timeout handling when using Limited Discoverable flag
**Describe the bug**
Issues in advertising timeout when application has set BT_LE_AD_LIMITED flag in advertising data.

- bt_le_ext_adv_stop() and bt_le_adv_stop() should also cancel delayed work ie. if called by application before timeout is reached
- delayed work should be cancelled when set terminated ie due to connection or parameters set via BT_LE_EXT_ADV_START_PARAM
- if advertising is terminated, how application will know about this? should "sent" callback be called? what about legacy advertising (with ext_adv disabled in kconfig)?

**Impact**
Using the BT_LE_AD_LIMITED flag leaves the advertising set timeout running, this could potentially stop an advertiser it was not supposed to stop, or lead to an ASSERT  (if CONFIG_ASSERT is enabled).

Discussion: https://github.com/zephyrproject-rtos/zephyr/pull/37426#discussion_r682509680

### PR DESCRIPTION
**Pull request by KKopyscinski**
Advertising might stop when:
- it was stopped by application
- device connected to a peer
- extended advertising reached stop condition
  defined in BT_LE_EXT_ADV_START_PARAM - this is handled in ll

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

Fixes #37467

### CODE REVIEW

**Comment by KKopyscinski** [comment]
FYI @sjanc

File: subsys/bluetooth/host/adv.c [hunk 1]
```diff
@@ -997,6 +997,23 @@ int bt_le_adv_start_ext(struct bt_le_ext_adv *adv,
 
 static void adv_timeout(struct k_work *work);
 
+void bt_le_lim_adv_cancel_timeout(struct bt_le_ext_adv *adv)
+{
+	int rc;
+
+	rc = k_work_delayable_remaining_get(&adv->timeout_work);
```
Comment (joerchan) [inline subsys/bluetooth/host/adv.c:1004]: This is not safe, there might be no remaining time i.e timeout is out, but work is still not executed.
I suggest you just cancel always. That's the recommended approach.

**Review by joerchan — CHANGES_REQUESTED** [review]
Unsafe cancel of delayed work.

File: subsys/bluetooth/host/adv.c [hunk 2]
```diff
@@ -997,6 +997,16 @@ int bt_le_adv_start_ext(struct bt_le_ext_adv *adv,
...
 static void adv_timeout(struct k_work *work);
 
+void bt_le_lim_adv_cancel_timeout(struct bt_le_ext_adv *adv)
+{
+	int rc;
+
+	(void)k_work_cancel_delayable(&adv->timeout_work);
+	rc = k_work_delayable_busy_get(&adv->timeout_work) != 0;
+	/* At this point timeout must be canceled*/
+	__ASSERT(rc == 0, "Failed to stop limited advertising timeout");
```
Comment (trond-snekvik) [inline subsys/bluetooth/host/adv.c:1007]: IMO, it'd be better to use a flag for this than asserting. I'm not confident that this couldn't fail.

The most common pattern for handling cancel failures is to set a flag right after the timer is started and checking it in the beginning of the timeout handler. In this cancel function, you can clear the flag and cancel the delayable. If the cancel call fails, the flag will prevent misfires.

File: subsys/bluetooth/host/adv.c [hunk 3]
```diff
@@ -1360,18 +1373,22 @@ int bt_le_ext_adv_delete(struct bt_le_ext_adv *adv)
...
+
 	struct k_work_delayable *dwork;
 	struct bt_le_ext_adv *adv;
 
 	dwork = k_work_delayable_from_work(work);
 	adv = CONTAINER_OF(dwork, struct bt_le_ext_adv, timeout_work);
 
-	if (atomic_test_bit(adv->flags, BT_ADV_EXT_ADV)) {
+#if defined(CONFIG_BT_EXT_ADV)
+	if (atomic_test_bit(adv->flags, BT_ADV_EXT_ADV) &&
```
Comment (joerchan) [inline subsys/bluetooth/host/adv.c:1384]: This check is incorrect. You should call bt_le_adv_stop() if `adv == bt_dev.adv` (Meaning it is the legacy advertiser and was started with bt_le_adv_start()), otherwise use bt_le_ext_adv_stop() (because it was started with bt_le_ext_adv_start()).

Please make this separate commit as it is a different problem than the timeout fix.

File: subsys/bluetooth/host/adv.c [hunk 4]
```diff
@@ -997,6 +997,14 @@ int bt_le_adv_start_ext(struct bt_le_ext_adv *adv,
 
 static void adv_timeout(struct k_work *work);
 
+void bt_le_lim_adv_cancel_timeout(struct bt_le_ext_adv *adv)
+{
+	/* Unset flag to disable timeout in case of work cancel failure */
+	adv->timeout_work_active = false;
+	(void)k_work_cancel_delayable(&adv->timeout_work);
+	k_work_delayable_busy_get(&adv->timeout_work);
```
Comment (joerchan) [inline subsys/bluetooth/host/adv.c:1005]: The busy status is not used, so this should be removed.

File: subsys/bluetooth/host/adv.c [hunk 5]
```diff
@@ -1360,18 +1373,22 @@ int bt_le_ext_adv_delete(struct bt_le_ext_adv *adv)
...
+		adv->timeout_work_active) {
 		err = bt_le_ext_adv_stop(adv);
 	}
 #else
-	err = bt_le_adv_stop();
+	if (adv->timeout_work_active) {
+		err = bt_le_adv_stop();
+	}
 #endif
 	__ASSERT(err != 0, "Limited Advertising timeout reached, "
```
Comment (joerchan) [inline subsys/bluetooth/host/adv.c:1393]: The assert seems like a bad idea. You'd have to be 100% sure that you don't have any race on who's stopping the advertiser.
And failing to stop and advertiser (because it is already stopped) isn't something that should warrant an assert in my opinion. I would feel much better if this was a BT_WARN instead.

File: subsys/bluetooth/host/hci_core.h [hunk 6]
```diff
@@ -152,6 +152,8 @@ struct bt_le_ext_adv {
 #endif /* defined(CONFIG_BT_EXT_ADV) */
 
 	struct k_work_delayable	timeout_work;
+
+	bool timeout_work_active;
```
Comment (trond-snekvik) [inline subsys/bluetooth/host/hci_core.h:156]: Could this be added as a bit in `adv->flags` instead?

**Comment by carlescufi** [comment]
@KKopyscinski can you please address the review comments?

**Comment by joerchan** [comment]
@sjanc What's the reason for closing this pull request?
